programs/scope/src/oracles/redstone.rs

```diff
@@ -0,0 +1,94 @@
+use anchor_lang::prelude::*;
+
+use crate::{utils::account_deserialize, warn, DatedPrice, Price, ScopeError};
+
+#[cfg(not(feature = "skip_price_validation"))]
+/// Price is kept in 0.05% deviation threshold all the time.
+/// At least one update per 30h will happen.
+const VALID_PRICE_LIFE_TIME: i64 = 30 * crate::utils::SECONDS_PER_HOUR;
+
+const RESERVED_BYTE_SIZE: usize = 64;
+
+#[account]
+struct PriceData {
+    pub feed_id: [u8; 32],
+    pub value: [u8; 32],
+    pub timestamp: u64,
+    pub write_timestamp: Option<u64>,
+    pub write_slot_number: u64,
+    pub decimals: u8,
+    pub _reserved: [u8; RESERVED_BYTE_SIZE],
+}
+
+fn redstone_value_to_price(raw_be_value: [u8; 32], decimals: u8) -> Result<Price> {
+    if !raw_be_value.iter().take(24).all(|&v| v == 0) {
+        warn!("Price overflow u64");
+        return Err(ScopeError::PriceNotValid.into());
+    }
+
+    let value = u64::from_be_bytes(raw_be_value[24..].try_into().unwrap());
+
+    Ok(Price {
+        value,
+        exp: decimals as u64,
+    })
+}
+
+#[cfg(not(feature = "skip_price_validation"))]
+fn check_price_life_time(timestamp_ms: Option<u64>, clock: &Clock) -> Result<()> {
+    let timestamp_ms = match timestamp_ms {
+        Some(ts) => ts,
+        None => {
+            warn!("Price feed was not updated yet");
+            return Err(ScopeError::PriceNotValid.into());
+        }
+    };
+
+    let ms_since_last_udpate = (clock.unix_timestamp * 1_000).checked_div(timestamp_ms as i64);
+
+    match ms_since_last_udpate {
+        Some(ms) if ms <= VALID_PRICE_LIFE_TIME => Ok(()),
+        _ => {
+            warn!(
+                "Redstone price feed account data has not been refreshed for more than: {:?}ms",
+                VALID_PRICE_LIFE_TIME
+            );
+            return Err(ScopeError::PriceNotValid.into());
+        }
+    }
+}
+
+pub fn get_price(price_info: &AccountInfo, _clock: &Clock) -> Result<DatedPrice> {
+    let price_data: PriceData = account_deserialize(price_info)?;
+    let price = redstone_value_to_price(price_data.value, price_data.decimals)?;
+
+    #[cfg(not(feature = "skip_price_validation"))]
+    check_price_life_time(price_data.write_timestamp, _clock)?;
+
+    Ok(DatedPrice {
+        price,
+        last_updated_slot: price_data.write_slot_number,
+        unix_timestamp: price_data
+            .write_timestamp
+            .expect("Checked in `check_price_life_time`")
+            / 1_000,
+        generic_data: Default::default(),
+    })
+}
+
+pub fn validate_account(price_data: &Option<AccountInfo>) -> Result<()> {
+    if cfg!(feature = "skip_price_validation") {
```

Comment:
is that the same cfg flag as for price_validation?

Reply:
yes